### PR TITLE
Add custom hooks and refactor store

### PR DESCRIPTION
**Pull request by alpanayotov**
Use custom hooks as per [Redux recommendation](https://react-redux.js.org/using-react-redux/usage-with-typescript) and refactor the store structure.

### CODE REVIEW

File: src/components/footer/index.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { useDispatch } from 'react-redux';
 import { useTranslation } from 'react-i18next';
 
 import { i18n, locales } from '@i18n';
-import { AuthActionType } from '@store/auth';
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/components/footer/index.tsx:7]: Shouldn\t this be exported from the `auth` branch?

File: src/components/header/index.tsx [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
-import { useSelector, useDispatch } from 'react-redux';
+import { useDispatch } from 'react-redux';
 
 import { Routes } from '@utilities';
+import { history } from '@store/index';
+import { AuthState } from '@store/interfaces';
 import { Icon, Button } from '@components';
-import { history, RootStore } from '@store/index';
-import { AuthState, AuthActionType } from '@store/auth';
+import { useAppSelector } from '@src/utilities/hooks';
```
Comment (scriptex) [inline src/components/header/index.tsx:10]: Please remove `src/`

File: src/components/header/index.tsx [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/components/header/index.tsx:11]: I think this should be exported from the `auth` branch.

File: src/components/login-form/index.test.tsx [hunk 4]
```diff
@@ -4,8 +4,11 @@ import { shallow } from 'enzyme';
 import { LoginForm } from '.';
 
 jest.mock('react-redux', () => ({
-	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(() => ({
+	connect: (): jest.Mock => jest.fn()
+}));
+
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/components/login-form/index.test.tsx:10]: You can omit the `src/` here too.

File: src/components/header/index.tsx [hunk 5]
```diff
@@ -1,19 +1,21 @@
...
 
 import './index.scss';
 
 export const Header: React.FunctionComponent = () => {
 	const { t } = useTranslation();
 	const dispatch = useDispatch();
-	const authState: AuthState = useSelector((store: RootStore) => store.auth);
+	const authState: AuthState = useAppSelector(state => state.auth);
```
Comment (scriptex) [inline src/components/header/index.tsx:18]: Isn't TypeScript inferring the type here?

File: src/components/login-form/index.tsx [hunk 6]
```diff
@@ -1,10 +1,9 @@
 import * as React from 'react';
 import { useForm } from 'react-hook-form';
-import { useSelector } from 'react-redux';
 import { useTranslation } from 'react-i18next';
 
-import { RootStore } from '@store/index';
 import { Field, Button } from '@components';
+import { useAppSelector } from '@src/utilities/hooks';
```
Comment (scriptex) [inline src/components/login-form/index.tsx:6]: `src/`

File: src/components/password-reset-form/index.tsx [hunk 7]
```diff
@@ -1,10 +1,9 @@
 import * as React from 'react';
 import { useForm } from 'react-hook-form';
-import { useSelector } from 'react-redux';
 import { useTranslation } from 'react-i18next';
 
-import { RootStore } from '@store/index';
 import { Field, Button } from '@components';
+import { useAppSelector } from '@src/utilities/hooks';
```
Comment (scriptex) [inline src/components/password-reset-form/index.tsx:6]: `src/`

File: src/components/password-reset-form/index.test.tsx [hunk 8]
```diff
@@ -4,8 +4,11 @@ import { shallow } from 'enzyme';
 import { PasswordResetForm } from '.';
 
 jest.mock('react-redux', () => ({
-	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(() => ({
+	connect: (): jest.Mock => jest.fn()
+}));
+
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/components/password-reset-form/index.test.tsx:10]: `src/`

File: src/components/signup-form/index.test.tsx [hunk 9]
```diff
@@ -4,8 +4,11 @@ import { shallow } from 'enzyme';
 import { SignupForm } from '.';
 
 jest.mock('react-redux', () => ({
-	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(() => ({
+	connect: (): jest.Mock => jest.fn()
+}));
+
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/components/signup-form/index.test.tsx:10]: `src/`

File: src/components/signup-form/index.tsx [hunk 10]
```diff
@@ -1,10 +1,9 @@
 import * as React from 'react';
 import { useForm } from 'react-hook-form';
-import { useSelector } from 'react-redux';
 import { useTranslation } from 'react-i18next';
 
-import { RootStore } from '@store/index';
 import { Field, Button } from '@components';
+import { useAppSelector } from '@src/utilities/hooks';
```
Comment (scriptex) [inline src/components/signup-form/index.tsx:6]: `src/`

File: src/containers/auth/login/index.test.tsx [hunk 11]
```diff
@@ -5,10 +5,13 @@ import { Login } from '.';
 
 jest.mock('react-redux', () => ({
 	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(),
 	useDispatch: (): jest.Mock => jest.fn()
 }));
 
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/containers/auth/login/index.test.tsx:11]: `src/`

File: src/containers/auth/login/index.tsx [hunk 12]
```diff
@@ -5,7 +5,7 @@ import { useTranslation } from 'react-i18next';
 
 import { Routes } from '@utilities';
 import { history } from '@store/index';
-import { AuthActionType } from '@store/auth';
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/containers/auth/login/index.tsx:8]: Again, should be exported from the `auth` branch :)

File: src/containers/auth/password-reset/index.test.tsx [hunk 13]
```diff
@@ -5,10 +5,13 @@ import { PasswordReset } from '.';
 
 jest.mock('react-redux', () => ({
 	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(),
 	useDispatch: (): jest.Mock => jest.fn()
 }));
 
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/containers/auth/password-reset/index.test.tsx:11]: `src/`

File: src/containers/auth/password-reset/index.tsx [hunk 14]
```diff
@@ -5,7 +5,7 @@ import { useTranslation } from 'react-i18next';
 
 import { Routes } from '@utilities';
 import { history } from '@store/index';
-import { AuthActionType } from '@store/auth';
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/containers/auth/password-reset/index.tsx:8]: `auth` branch

File: src/containers/auth/signup/index.test.tsx [hunk 15]
```diff
@@ -5,10 +5,13 @@ import { Signup } from '.';
 
 jest.mock('react-redux', () => ({
 	connect: (): jest.Mock => jest.fn(),
-	useSelector: jest.fn(),
 	useDispatch: (): jest.Mock => jest.fn()
 }));
 
+jest.mock('@src/utilities/hooks', () => ({
```
Comment (scriptex) [inline src/containers/auth/signup/index.test.tsx:11]: `src/`

File: src/containers/auth/signup/index.tsx [hunk 16]
```diff
@@ -5,7 +5,7 @@ import { useTranslation } from 'react-i18next';
 
 import { Routes } from '@utilities';
 import { history } from '@store/index';
-import { AuthActionType } from '@store/auth';
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/containers/auth/signup/index.tsx:8]: `auth` branch

File: src/index.tsx [hunk 17]
```diff
@@ -1,15 +1,14 @@
...
-import { Store } from 'redux';
 import { render } from 'react-dom';
 import { Provider } from 'react-redux';
 import { ConnectedRouter } from 'connected-react-router';
 
 import { App } from './app';
 import { removeItems } from '@utilities';
-import { AuthActionType } from '@store/auth';
-import { history, RootStore, configureStore } from '@store/index';
+import { AuthActionType } from '@store/enums';
```
Comment (scriptex) [inline src/index.tsx:8]: `auth` branch

File: src/store/enums.ts [hunk 18]
```diff
@@ -0,0 +1 @@
+export * from './branches/auth/enums';
```
Comment (scriptex) [inline src/store/enums.ts:1]: Oh, so you've taken a different approach here? Why don't we apply it to the rest of the files - `interfaces.ts` too for example?

File: src/utilities/hooks.ts [hunk 19]
```diff
@@ -0,0 +1,5 @@
+import { TypedUseSelectorHook, useSelector } from 'react-redux';
+
+import { RootState } from '@store/interfaces';
+
+export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
```
Comment (scriptex) [inline src/utilities/hooks.ts:5]: Do you want us to use the `useDispatch` the way it is explained in the [Redux docs](https://react-redux.js.org/using-react-redux/usage-with-typescript#define-typed-hooks)?

**Review by scriptex — CHANGES_REQUESTED** [review]

File: src/components/footer/index.tsx (hunk 1 above)
Comment (alpanayotov) [reply src/components/footer/index.tsx:7]: No, all enums and interfaces from the store branches have single point for import from `@store`

File: src/components/header/index.tsx (hunk 3 above)
Comment (alpanayotov) [reply src/components/header/index.tsx:11]: No, all enums and interfaces from the store branches have single point for import from @store

File: src/containers/auth/login/index.tsx (hunk 12 above)
Comment (alpanayotov) [reply src/containers/auth/login/index.tsx:8]: nope

File: src/containers/auth/password-reset/index.tsx (hunk 14 above)
Comment (alpanayotov) [reply src/containers/auth/password-reset/index.tsx:8]: nope

File: src/store/enums.ts (hunk 18 above)
Comment (alpanayotov) [reply src/store/enums.ts:1]: It's applied, it's just we don't import the interfaces outside the store on this phase

File: src/utilities/hooks.ts (hunk 19 above)
Comment (alpanayotov) [reply src/utilities/hooks.ts:5]: Let's do this in another PR. It requires a bit more effort than the docs are indicating and I want to ship the new store structure.

**Review by scriptex — APPROVED** [review]

**Pull request merged**